File: 152/152_version1.java
class Solution {
    public int maxProduct(int[] nums) {
        if(nums.length==0 || nums==null) return 0;
        int result = nums[0],minpro = nums[0],maxpro = nums[0];
        for(int index = 1;index < nums.length;index++){
            int value = nums[index];
            int temp = maxpro;
            maxpro = Math.max(Math.max(minpro*value,maxpro*value),value);
            minpro = Math.min(Math.min(minpro*value,temp*value),value);
            result = Math.max(maxpro,result);
        }
        return result;
    }
}

File: 152/152_version1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        maxproduct = nums[0]
        minproduct = nums[0]
        result = nums[0]
        index = 1
        while index < len(nums):
            num = nums[index]
            if num < 0:
                maxproduct,minproduct = minproduct,maxproduct
            maxproduct = max([num,maxproduct*num])
            minproduct = min([num,minproduct*num])
            result = max([result,maxproduct])
            index+=1
        return result     
